docs/071-BDD.md

```diff
@@ -0,0 +1,550 @@
+
+The idea of story BDD can be narrowed to:
+
+* describe features in a scenario with a formal text
+* use examples to make abstract things concrete
+* implement each step of a scenario for testing
+* write actual code implementing the feature
+
+By making every feature to be formally written (in User Story format) and test automated, we ensure that business, developers, QAs, managers are in the same boat. That everyone in a team knows what is developed, what is not, what is tested and what is not. By making tests to be a part of user story BDD allows that non-technical personnel to write (or edit) acceptance tests. BDD also encourages exploration and discussions. To formalize the requirements and write user story in a way everyone can understand them, business and developers should talk, ask questions, and play the scenarios of features to implement.
+
+### Ubiquitous Language
+
+The ubiquitous language is always referred as *common* language. That is it's main benefit. It is not a couple of our business specification's words, and not a couple of developer's technical terms. It is a common words and terms that can be understood by people for whom we are building the software and should be understood by developers. Establishing correct communication between this two groups people is vital for building successful project that will fit the domain and fulfill all business needs.
+
+Each feature of a product should be born from a talk between
+
+* business (analysts, product owner)
+* developers
+* QAs
+
+which are known in BDD as "three amigos".
+
+Such talks should produce written stories. There should be an actor that doing some things, the feature that should be fulfilled within the story and the result achieved. 
+
+We can try to write such simple story:
+
+```
+As a customer i want to buy several products
+I put first product with 600 $ price to my cart
+And then another one with 1000 $ price
+When i go to checkout process
+I should see that total number of products i want to by is 2
+And my order amount is 1600 $
+```
+
+As we can see this simple story highlights core concepts that are called *contracts*. We should fulfill those contracts to model software correctly. But how we can verify that those contracts are being satisfied? [Cucumber](http://cucumber.io) introduced a special language for such stories called **Gherkin**. Same story transformed to Gherkin will look like this:
+
+```gherkin
+Feature: checkout process
+  In order to buy products
+  As a customer
+  I want to be able to buy several products 
+
+Scenario:
+  Given I have product with 600 $ price in my cart 
+  And I have product with 1000 $ price
+  When I go to checkout process 
+  Then I should see that total number of products is 2
+  And my order amount is 1600 $
+```
+
+Cucucmber, Behat, and sure, **Codeception** can execute this scenario step by step as an automated test. 
+Every step in this scenario requires a code which defines .
```

Comment:
`... requires a code which defines .`

I don't know how to write it better, but the current sentence doesn't sound correct to me.